nav2_map_server/src/map_3d/map_saver_3d.cpp

```diff
@@ -0,0 +1,192 @@
+#include <stdexcept>
+#include <functional>
+
+#include "sensor_msgs/msg/point_cloud2.hpp"
+
+using namespace std::placeholders;
+
+namespace nav2_map_server
+{
+
+MapSaver3D::MapSaver3D()
+: nav2_util::LifecycleNode("map_saver", "", true)
+{
+  RCLCPP_INFO(get_logger(), "Creating");
+
+  save_map_timeout_ = std::make_shared<rclcpp::Duration>(
+    rclcpp::Duration::from_seconds(declare_parameter("save_map_timeout", 2.0)));
+
+  map_subscribe_transient_local_ = declare_parameter("map_subscribe_transient_local", true);
+}
+
+MapSaver3D::~MapSaver3D()
+{
+}
+
+nav2_util::CallbackReturn
+MapSaver3D::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+
+  // Make name prefix for services
+  const std::string service_prefix = get_name() + std::string("/");
+
+  // Create a service that saves the occupancy grid or PointCloud2 from map topic to a file
+  save_map_service_ = create_service<nav2_msgs::srv::SaveMap3D>(
+    service_prefix + save_map_service_name_,
+    std::bind(&MapSaver3D::saveMapCallback, this, _1, _2, _3));
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+MapSaver3D::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+
+  // create bond connection
+  createBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+MapSaver3D::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  // destroy bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+MapSaver3D::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  save_map_service_.reset();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+MapSaver3D::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void MapSaver3D::saveMapCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_msgs::srv::SaveMap3D::Request> request,
+  std::shared_ptr<nav2_msgs::srv::SaveMap3D::Response> response)
+{
+  map_3d::SaveParameters save_parameters;
+  save_parameters.map_file_name = request->map_url;
+  save_parameters.as_binary = request->as_binary;
+  save_parameters.format = request->file_format;
+
+  response->result = saveMapTopicToFile(request->map_topic, save_parameters);
+}
+
+bool MapSaver3D::saveMapTopicToFile(
+  const std::string & map_topic,
+  const map_3d::SaveParameters & save_parameters)
+{
+  // Local copies of map_topic and save_parameters that could be changed
+  std::string map_topic_loc = map_topic;
+  map_3d::SaveParameters save_parameters_loc = save_parameters;
+
+  RCLCPP_INFO(
+    get_logger(), "Saving map from \'%s\' topic and default origin to \'%s\' file",
```

Comment:
Not clear, what "default origin" means here: topic or `(0,0,0)` origin by-default?